projects/src/simbicycle.hh

```diff
@@ -146,35 +133,38 @@ void Bicycle<Model, Observer>::update_kinematics() {
 
     // mod rear wheel angle so it does not grow beyond all bounds
     model_t::set_full_state_element(m_full_state, full_state_index_t::rear_wheel_angle,
-            std::fmod(model_t::get_full_state_element(x, full_state_index_t::rear_wheel_angle),
+            std::fmod(model_t::get_full_state_element(full_state_copy, full_state_index_t::rear_wheel_angle),
                       constants::two_pi)),
     chBSemSignal(&m_state_sem);
 
     m_pose.timestamp = chVTGetSystemTime();
-    m_pose.x = model_t::get_full_state_element(x, full_state_index_t::x);
-    m_pose.y = model_t::get_full_state_element(x, full_state_index_t::y);
-    m_pose.rear_wheel = model_t::get_full_state_element(x, full_state_index_t::rear_wheel_angle);
+    m_pose.x = model_t::get_full_state_element(full_state_copy, full_state_index_t::x);
+    m_pose.y = model_t::get_full_state_element(full_state_copy, full_state_index_t::y);
+    m_pose.rear_wheel = model_t::get_full_state_element(full_state_copy, full_state_index_t::rear_wheel_angle);
     m_pose.pitch = pitch;
-    m_pose.yaw = model_t::get_full_state_element(x, full_state_index_t::yaw_angle);
+    m_pose.yaw = model_t::get_full_state_element(full_state_copy, full_state_index_t::yaw_angle);
     m_pose.roll = roll;
     m_pose.steer = steer;
 }
 
 template <typename Model, typename Observer>
-void Bicycle<Model, Observer>::prime_observer() {
-    if (std::is_same<Observer, typename observer::Kalman<Model>>::value) {
-        // define a non-zero initial state, this is selected arbitrarily for now
-        state_t x = state_t::Zero();
-        model::Bicycle::set_state_element(x, model_t::state_index_t::steer_angle, 0.1f); // in radians
-
-        float t = 0;
-        while (t < observer_prime_period) {
-            x = m_model.update_state(x);
-            m_observer.update_state(input_t::Zero(), m_model.calculate_output(x));
-            t += m_model.dt();
-        }
+OBSERVER_FUNCTION(void) Bicycle<Model, Observer>::prime_observer() {
```

Comment:
T is used force template argument deduction
https://stackoverflow.com/questions/30953248/why-doesnt-sfinae-enable-if-work-for-member-functions-of-a-class-template
And details of deduction are here:
http://en.cppreference.com/w/cpp/language/template_argument_deduction
although I haven't read through it throughly.